vowpalwabbit/core/src/reductions/gd.cc

```diff
@@ -825,7 +852,12 @@ void NO_SANITIZE_UNDEFINED learn(VW::reductions::gd& g, base_learner& base, VW::
   assert(ec.l.simple.label != FLT_MAX);
   assert(ec.weight > 0.);
   g.predict(g, base, ec);
+  // this state should only matter on learn and not predict
+  // assert(g.current_model_state == nullptr);
+  g.current_model_state = &(g.per_model_states.at(ec.ft_offset / g.all->weights.stride()));
```

Comment:
I think we need to use the VW::scope_exit(

Reply:
done